### PR TITLE
New blog post about sandboxing JS execution.

### PR DESCRIPTION
**Pull request by malaporte**

### CODE REVIEW

File: _posts/2015-04-26-sandboxing-javascript-execution-in-java.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+To run JS code with quota checks, you need to run the code using the `ExecutionContext` returned by `createResourceQuotaExecutionObject` on an existing `ExecutionContext`. It is possible to invoke several blocks of code using the same quotas. The total consumed resources by all script executions will be used to check the quotas.
+
+{% highlight java %}
+context.createResourceQuotaExecutionObject(new ResourceQuota(cpuQuota, memoryQuota));
+{% endhighlight %}
+
+## What about Nashorn?
+
+[Nashorn](http://openjdk.java.net/projects/nashorn/) is the new JS interpreter bundled with Java 8. It has good performance and is certainly more robust, but I still haven't figured out a way to implement proper CPU and memory quotas in that engine, mainly because I haven't yet find a place where I can regularly check if quotas have been exceeded (I haven't tried very hard though). I might write a new post when/if I succeed in that endeavour.
```
Comment (wfortin) [inline _posts/2015-04-26-sandboxing-javascript-execution-in-java.md]: I haven't yet `found`

File: _posts/2015-04-26-sandboxing-javascript-execution-in-java.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  image: mlaporte.jpg
+---
+
+The [Query Pipeline](https://developers.coveo.com/display/SearchREST/Managing+the+Query+Execution+Pipeline) on a Coveo index can be extended using JavaScript code running on the server. This has many benefits, but allowing arbitrary code to run in a server process opens the possibility for an admin to bring the server down in many creative ways. To protect against this we added functionality to the JS interpreter we're using ([DynJS](https://github.com/dynjs/dynjs)) to prevent Bad Things from happening.
+
+<!-- more -->
+
+There are three ways external code running in a process can cause harm:
+
+* Using APIs to damage the server process or underlying server
```
Comment (jebeaudet) [inline _posts/2015-04-26-sandboxing-javascript-execution-in-java.md]: server processes ?

File: _posts/2015-04-26-sandboxing-javascript-execution-in-java.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    long delta = current - lastCpuTime;
+    lastCpuTime = current;
+
+    if (cpuTime.addAndGet(-delta) < 0) {
+        throw new ExceededResourcesException("CPU usage has exceeded the allowed quota");
+    }
+}
+{% endhighlight %}
+
+A similar technique is used to monitor heap allocations. The same `ThreadMXBean` object also exposes metrics about how many bytes were allocated by the current thread (including memory that is no longer referenced, but's that OK). By checking this metric in the exact same way as for CPU, we can detect whenever the thread has exceeded the allowed memory quota and put an end to it's processing.
```
Comment (jebeaudet) [inline _posts/2015-04-26-sandboxing-javascript-execution-in-java.md]: put an end to its processing (no `'`

File: _posts/2015-04-26-sandboxing-javascript-execution-in-java.md (hunk 2 above)
Comment (malaporte) [inline _posts/2015-04-26-sandboxing-javascript-execution-in-java.md]: Euh, no I think.

Comment (jebeaudet) [inline _posts/2015-04-26-sandboxing-javascript-execution-in-java.md]: Didn't you mean all the server processes? It that case it's plural. Unless you meant only the JVM process then nevermind me :+1:

**Pull request merged**